example-app/src/main/java/com/adyen/checkout/example/ui/googlepay/compose/SessionsGooglePayScreen.kt

```diff
@@ -50,12 +46,17 @@ import com.google.pay.button.PayButton
 @Composable
 internal fun SessionsGooglePayScreen(
     useDarkTheme: Boolean,
+    googlePayState: SessionsGooglePayState,
+    eventsState: SessionsGooglePayEvents,
+    onGooglePayLauncherResult: (apiTaskResult: ApiTaskResult<PaymentData>) -> Unit,
     onBackPressed: () -> Unit,
-    viewModel: SessionsGooglePayViewModel,
 ) {
+    lateinit var googlePayComponent: GooglePayComponent
```

Comment:
Another question, is there any case where this field might not get initialized and causes a problem? Asking because you definitely know more than me about compose 😅

Reply:
That's a question for us, since the component is only accessed in response to events and state changes under our control. In other words, if we can guarantee that the `fetchSession` action (triggering the `SessionsGooglePayEvents.ComponentData` event) completes before any of the following happens:
- A `SessionsGooglePayEvents.Action` or `SessionsGooglePayEvents.Intent` event is issued.
- The pay button is clicked
- The state becomes `SessionsGooglePayState.ShowButton` or `SessionsGooglePayState.ShowComponent`.

...then the logic must be safe.
My understanding is that this is the case at least for all Google Pay related events, since Google Pay availability is only requested one the session call returns. Is that also true for `Action`, or `Intent` events sent by the component?

On that note, I see that we are handling the `SessionsGooglePayState.ShowComponent` state in the compose UI, but that value is never set elsewhere.

Reply:
The var is function local, so access modifiers are not necessary.

Reply:
The tricky part here is with `Intent`. This is usually triggered after a redirect (custom tabs, external browser, app switch) has completed and we returned to the app through a deep link. With Google Pay that's mostly after the Google Pay component is finished and a 3DS challenge was launched in the browser (which is rare).

Right now we have this code in the `SessionsGooglePayActivity`:

```kotlin 
    override fun onNewIntent(intent: Intent) {
        ..
        sessionsGooglePayViewModel.onNewIntent(intent) // this triggers a `SessionsGooglePayEvents.Intent`
    }
```

The only case I can think of where this code executes before the `SessionsGooglePayEvents.ComponentData` event has been called is if the activity got destroyed in the background.

But then even without a redirect, I don't think this sample covers that scenario (app destroyed in the background and recreated), because we are fetching a new session when the view model is created so we lose the previous state of the component completely. I am not sure if any of our samples covers that use case, it's a tricky one to deal with.

Reply:
For compose, note that the view model can outlive the activity (and hence the data there can be reused in subsequent activity recreations).

If the 3DS flow is added to the app, as a developer, I'd like to see one of the following (in order of preference):
- A good practice sample that I can use in my app (including the session renewal)
- A marked commented warning letting developers know about lifecycle challenges

If this is not deemed relevant for a simple sample application, I'd consider removing it otherwise.

Reply:
Thanks for the input! I think we definitely would like to provide an example that fully works with all edge case scenarios in mind.

> and hence the data there can be reused in subsequent activity recreations

Yes and in this case our samples would work. However (correct me if I'm wrong) if the system decides to kill the app in the background to free up memory (just like when you switch on "Don't Keep Activities" in the developer options) then everything in memory will be destroyed including the view model. Which means the only thing we can use to recover the data is the `SaveStateHandle`. We already use it on the SDK side but not for all scenarios, so we need to investigate this more.